src/coreclr/vm/dynamicmethod.cpp

```diff
@@ -1475,6 +1475,44 @@ void LCGMethodResolver::GetEHInfo(unsigned EHnumber, CORINFO_EH_CLAUSE* clause)
 
 #endif // !DACCESS_COMPILE
 
+bool DynamicMethodDesc::HasMDContextArg()
+{
+    CONTRACTL
+    {
+        MODE_ANY;
+        GC_NOTRIGGER;
+        NOTHROW;
+        PRECONDITION(IsILStub());
+    }
+    CONTRACTL_END;
+
+    // Perform a minimal check. This is historically sufficient, but has not been updated
+    // with flag usage.
+    // If the check indicates false, it is for sure false. However, if it is true
+    // we need to check additional cases.
+    bool minCheck = IsCLRToCOMStub() || (IsPInvokeStub() && !IsDelegateStub());
+    if (!minCheck)
+        return false;
+
+#ifdef DACCESS_COMPILE
+    // The DAC's usage of this API is narrow enough that the precise nature needed by
+    // the runtime is not needed. Therefore, we can return the historically sufficient answer.
+    return true;
+#else
+    ILStubResolver::ILStubType type = GetILStubResolver()->GetStubType();
```

Comment:
What is this actually needed for? I would expect that the `minCheck` covers all cases.

Reply:
Ideally it would, but there is something missing for the ArrayOp stubs. The extended flags for the `MethodDesc` are set in such a way this permits true for the stub.

```c++
    bool IsCLRToCOMStub()
    {
        LIMITED_METHOD_CONTRACT;
        _ASSERTE(IsILStub());
        return ((0 == (m_dwExtendedFlags & mdStatic))
            && !IsReverseStub()
            && !IsDelegateStub()
            && !IsStructMarshalStub());
    }
```

Reply:
Instead of adding yet another flag for array, I decided to just be explicit about the stubs that have a `MethodDesc`.

Reply:
If `IsCLRToCOMStub` does not return what its name says, we should rather fix that. 

I think it is a good idea to have enum on the IL stub MethodDesc that says what kind of stub it represents. We have an enum like on IL stub resolver that is initialized using `pMD->GetILStubResolver()->SetStubType(ILStubResolver::...)`. Should we move this stub type enum to the IL stub MethodDesc so that it is always available?

Reply:
>If IsCLRToCOMStub does not return what its name says, we should rather fix that.

Agreed, but that was proving to be difficult so I commented the state and hid it under the `HasMDContextArg()` so fixing it in the future would be relatively simple.

> I think it is a good idea to have enum on the IL stub MethodDesc that says what kind of stub it represents. We have an enum like on IL stub resolver that is initialized using pMD->GetILStubResolver()->SetStubType(ILStubResolver::...). Should we move this stub type enum to the IL stub MethodDesc so that it is always available?

So that is definitely interesting and something we could explore. My concern there is the amount of bits. we don't have the space for all this data - at least not in its current form. I can take a bit and see if I can get it into the `MethodDesc` but that is likely a fair bit of churn in an area we likely have little testing.

Reply:
These bits need to be on DynamicMethodDesc, not MethodDesc. I do not think that it should be particularly hard to encode store the full kind on DynamicMethodDesc, without growing its size. There are some free bits in `DynamicMethodDesc:: ExtendedFlags` and the existing bits partially encode the stub kind already (e.g. nomdCALLIStub, nomdDelegateStub, nomdStructMarshalStub, ...). The existing encoding is wasting bits quite a bit by encoding the kind as a bit mask. You can save a lot of the existing bits by encoding it as enum.

Reply:
@jkotas Thanks for seeing the forest on this one. I took a closer look at `m_dwExtendedFlags`. I can make this a lot better.